File: backend/src/modules/reporting/svc.ts
import type { ParentDocument } from "./model.ts";
import { type IReportingRepo, type IS3Repository } from "./repo.ts";
import type { DocumentType } from "../../shared/types/reporting.ts";
import { extname } from "path";
import type { AsyncResponseType } from "../../shared/types/response.ts";

export interface IReportingSvc {
  getParentDocumentByType(
    userId: string,
    documentType: DocumentType,
  ): AsyncResponseType<ParentDocument>;
  getParentDocumentTypeByDocumentId(
    parentDocumentId: string,
  ): AsyncResponseType<DocumentType>;
  saveParentDocument(
    userId: string,
    documentType: DocumentType,
    file: File,
  ): AsyncResponseType<ParentDocument>;
  getDocumentURLByFilePath(
    bucket: string,
    key: string,
    expiresIn: number,
  ): AsyncResponseType<string>;
  getParentDocumentURLByDocumentID(docId: string): AsyncResponseType<string>;
}

export class ReportingSvc implements IReportingSvc {
  private repo: IReportingRepo;
  private s3Repo: IS3Repository;
  constructor(repo: IReportingRepo, s3Repo: IS3Repository) {
    this.repo = repo;
    this.s3Repo = s3Repo;
  }

  async getParentDocumentByType(
    userId: string,
    documentType: DocumentType,
  ): AsyncResponseType<ParentDocument> {
    return await this.repo.getParentDocumentByType(userId, documentType);
  }

  async getParentDocumentTypeByDocumentId(
    parentDocumentId: string,
  ): AsyncResponseType<DocumentType> {
    return await this.repo.getParentDocumentTypeByDocumentId(parentDocumentId);
  }

  async saveParentDocument(
    userId: string,
    documentType: DocumentType,
    file: File,
  ): AsyncResponseType<ParentDocument> {
    const filePath = `parents/${userId}/documents/${documentType}${extname(file.name)}`;
    const { error } = await this.s3Repo.uploadFile("mybucket", filePath, file);
    if (error) return { data: undefined, error };
    const result = await this.repo.saveParentDocument(
      userId,
      documentType,
      filePath,
    );
    if (result.error) return { data: undefined, error: result.error };
    return result;
  }

  async getDocumentURLByFilePath(
    key: string,
    bucket: string = "mybucket",
    expiresIn: number = 3600,
  ): AsyncResponseType<string> {
    return await this.s3Repo.getDocumentURLByFilePath(key, bucket, expiresIn);
  }

  async getParentDocumentURLByDocumentID(
    docId: string,
  ): AsyncResponseType<string> {
    const { data, error } =
      await this.repo.getParentDocumentFilePathByDocumentID(docId);
    if (error) return { data: undefined, error };
    return await this.s3Repo.getDocumentURLByFilePath(data);
  }
}

File: apps/backend-python/src/modules/reporting/svc.py
from abc import ABC, abstractmethod
from uuid import UUID
from src.modules.reporting.repo import IReportingRepo
from src.modules.reporting.s3 import IS3Repository
from src.shared.types.modules.reporting.enum import DOCUMENT_TYPE
from src.shared.types.modules.reporting.model import ParentDocument
from src.shared.types.response import HTTPErrorResponse


class IReportingSvc(ABC):
    @abstractmethod
    async def get_parent_document_by_type(
        self,
        user_id: UUID,
        document_type: DOCUMENT_TYPE,
    ) -> HTTPErrorResponse[ParentDocument]:
        pass

    @abstractmethod
    async def get_parent_document_url_by_document_id(
        self, doc_id: UUID
    ) -> HTTPErrorResponse[str]:
        pass

    @abstractmethod
    async def get_document_url_by_file_path(
        self,
        path: str,
    ) -> HTTPErrorResponse[str]:
        pass


class ReportingSvc(IReportingSvc):
    def __init__(self, repo: IReportingRepo, s3_repo: IS3Repository) -> None:
        self.repo = repo
        self.s3_repo = s3_repo

    async def get_parent_document_by_type(
        self,
        user_id: UUID,
        document_type: DOCUMENT_TYPE,
    ) -> HTTPErrorResponse[ParentDocument]:
        return await self.repo.get_parent_document_by_type(
            user_id=user_id, document_type=document_type
        )

    async def get_parent_document_url_by_document_id(
        self, doc_id: UUID
    ) -> HTTPErrorResponse[str]:
        path, error = await self.repo.get_parent_document_file_path_by_document_id(
            doc_id=doc_id
        )
        if error:
            return None, error
        assert path is not None
        return await self.get_document_url_by_file_path(path=path)

    async def get_document_url_by_file_path(
        self,
        path: str,
    ) -> HTTPErrorResponse[str]:
        return await self.s3_repo.get_document_url_by_file_path(key=path)
